refactor(spec-specs): refactor EIP-7928 to use nested context frames

fselmo opened the pull request:
## 🗒️ Description

Related to comment [here](https://github.com/ethereum/execution-specs/pull/1759#issuecomment-3492476620).

This is a rough first-pass implementation but one that is passing all EIP-7928 tests at the moment. If we decide to go with this approach, this is a good base to get the tests released and we can make tweaks to the design as future PRs on top of `eips/amsterdam/eip-7928`.

## ✅ Checklist
<!-- Please check off all required items. For those that don't apply remove them accordingly. -->

- [x] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx tox -e static
    ```
- [x] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.

#### Cute Animal Picture

<img width="679" height="610" alt="Screenshot 2025-11-06 at 09 07 26" src="https://github.com/user-attachments/assets/effa317e-f576-4805-a6ab-1f4cf4f7b810" />

fselmo committed:
fix(spec-specs): require and use blockenv for state tracking

fselmo committed:
refactor(spec-specs): track BAL changes via frames

SamWilsn reviewed (commented):
Again, sorry for the scattered and disorganized feedback!

## `block_access_lists/rlp_utils.py`

### `rlp_encode_block_access_list`

It looks like you've reimplemented the logic from [ethereum-rlp](https://github.com/ethereum/ethereum-rlp/blob/eab94d5400819fa7a738d705a5e6a19f9a7ec683/src/ethereum_rlp/rlp.py#L84-L85) (treat a dataclass as a list of its children). If the encoding algorithm is standard RLP, you can probably just use `encode` on the `BlockAccessList` itself. If it requires slight tweaking, you could look at [`With`](https://github.com/ethereum/ethereum-rlp/blob/eab94d5400819fa7a738d705a5e6a19f9a7ec683/tests/test_rlp.py#L454-L462). If you've already investigated those options, please disregard!

### `validate_block_access_list_against_execution`

Probably doesn't belong in this file, since it has little to do with RLP.

Also, seems unused?

## `block_access_lists/builder.py`

### `build_block_access_list`

You should use [`to_be_bytes32`](https://ethereum.github.io/ethereum-types/src/ethereum_types/numeric.py.html#ethereum_types.numeric.Unsigned.to_be_bytes32:0) to convert to `Bytes32`.

I haven't reached `StateChanges` yet, so this comment might be premature (and if it is, and you're reading this, I forgot to remove it), but what's different between a `BlockAccessListBuilder` and ` StateChanges`? They seem to have very similar fields.

## `block_access_lists/rlp_types.py`

These aren't specifically RLP types, though they do get encoded to RLP. I'd name the file something different.

I'm not sure adding the type aliases from EIP-7928 actually make anything more clear. If we do want to go down the road of making semantically meaningful types, we should do it more broadly and uniformly.

We can add `U16` to ethereum-types if you'd like.

The constants here seem out of place.

## `fork.py`

### `process_transaction`

It seems like you're constructing a second stack of frames. Unless there's a reason not to, I'd rather add a member to `TransactionEnvironment` .

Is `increment_index` replaceable with `TransactionEnvironment.index_in_block + 1` maybe? I'm not 100% sure how `index_in_block` interacts with system txs.

## `state.py`

Seems a bit odd that we're passing the `StateChanges` into these functions. If we're trying to cleanly separate the "storage of state" from "tracking state changes", we'd want to avoid mixing the two.

## `state_tracker.py`

Overall comment, we don't use methods (`def x(self, ...)`). Everything should be a free function.

### `StateChanges`

Should we introduce some named tuples or something? There are a lot of one and two element tuples here, and I feel like that could get confusing without human readable names.

How do the pre-state captures interact with reverts and nested frames and such? I'm assuming that these are captured pre-transaction, so it might make more sense to put this in `TransactionEnvironment`?

#### `get_block_access_index`

Should this be set when constructing the `StateChanges`, or can its value change and therefore needs to be recomputed?

#### `merge_on_success` / `merge_on_failure`

Would it be better if these asserted `self.parent is not None`? Is there a case where we call it where we don't know where we are in the stack?

### `handle_in_transaction_selfdestruct`

I'd probably just rename this to `track_selfdestruct` (or `track_self_destruct`, whichever is consistent with our naming). All self-destructs in recent forks are in the same tx, right?

---

And this is all I have time for today!

SamWilsn commented:
## `eoa_delegation.py`

### `check_delegation`

Probably want to make some kind of structure with named fields for the return type (and backport that to previous forks eventually).

### `apply_delegation_tracking`

Rename to `track_delegation` maybe? Seems more consistent with the other tracking functions.

### `access_delegation`

Noting only for completeness, but if we're deprecating this, we'd want to actually remove it instead.

## `environment.py`

- I'd swap `is_cold_access` for `is_warm_access`, so we can avoid the cognitive overhead of `not`.
 - These changes'll need to be backported to the other forks:
    ```patch
    - code = get_account(evm.message.block_env.state, address).code
    + state = evm.message.block_env.state
    + code = get_account(state, address).code
    ```

## `system.py`

### `generic_create`

Is this a bug in the existing forks, and if so, should the ordering change be made to all previous forks as well?

## `interpreter.py`

Can we avoid a parent reference in `StateChanges` itself? I'd _really_ prefer to avoid creating a second representation of the stack, instead relying on `Evm`.

### `create_call_frame`

I assume this is a leftover, or will there be more stuff in here eventually?

### `get_message_state_frame`

The name of this function is misleading since calling it multiple times would create separate `StateChanges` instances.

## `genesis.py`

### `add_genesis_block`

The type of `Block.block_access_list` is `BlockAccessList` but the return type of `rlp.encode` is `Bytes`.

---

## Miscellaneous

- Should we rename `Evm.accessed_addresses` to `Evm.warmed_addresses` to avoid creating confusion between two types of "accessed" addresses?
- Some comments describe the behaviour of distant code. If we ever update that distant code, we won't know to update these comments, leading to a possibly confusing situation. Not sure how common this is in our codebase, but now that I see it, maybe we should go through and look?

fselmo committed:
fix(spec-specs): Mark original addr warm before delegation

Co-authored-by: spencer spencer.taylor-brown@ethereum.org

fselmo committed:
fix(spec-specs): Make sure we account for no changes

fselmo committed:
fix(spec-specs): Better tracking for code changes; ensure with BAL test

fselmo committed:
fix(spec-specs): Use child frame for create message

fselmo committed:
fix(spec-specs): Normalize transaction before merging to block frame

fselmo committed:
fix(spec-specs): Early static check for SSTORE before any reads

fselmo committed:
fix(spec-specs): Track storage writes more appropriately wrt index

fselmo committed:
fix(spec-specs): Use functions, not methods; fix create revert

fselmo committed:
fix(spec-specs): Default code to b"" in tracker, skip empty setting

fselmo committed:
fix(spec-specs): Fix BAL cross-transaction tracking and nonce dedup

 - add commit_transaction_frame() - no net-zero filtering for cross-tx changes
 - keep max nonce per transaction when building BAL, remove block-level code filtering
 - filter net-zero code changes at tracking time (for 7702 txs)
 - use commit_transaction_frame() instead of merge_on_success() for tx->block commits

fselmo committed:
fix(spec-specs): Move destroy_account before BAL normalization

fselmo committed:
fix(spec-specs): Check delegation access gas before reading

fselmo committed:
fix(spec-specs): Track code per auth; filter pre at tx frame

fselmo committed:
fix(spec-specs): Use proper frames for system transactions
